### ISSUE TITLE
Copying the private key of the CA to everyhost is a security issue

### ISSUE THREAD

**Issue by j-lowry** [CiscoCloud/mantl#1766]
I am specifically looking at PR #1761 and looking at how mantl handles CAs and certificates. Giving each host a non-password protected copy of the CA key means any attacker that is able to get onto a host can issue themselves a valid certificate to authenticate with all the mantl services. There is a further issue that the attacker can also just grab the keypair that already exists and auth that way. Re-keying would be required and would knock everything offline in the meantime since nothing would be trusted.

Proposed solution, admittedly very rough and without an expert understanding of ansible:
 - create a root CA locally
 - for each host that is created, copy the CA public as we are currently doing
 - create a private key, csr, and then signed certificate locally, copying to each host the private/public key pair from the locally generated keypairs

Future improvements:
 - bootstrap consul and vault to get a keypairs signed by the root CA.
 - create an intermediate CA and put it into Hashicorp vault's PKI backend for other hosts after that.

I would love to help out to make this happen, just need some guidance.

- Ansible version (`ansible --version`):  ansible 2.0.1.0
- Python version (`python --version`): Python 2.7.10
- Git commit hash or branch: master
- Cloud Environment: AWS
- Terraform version (`terraform version`): Terraform v0.6.15

**Comment by siddharthist** [comment]
@j-lowry Thanks for bringing this up! This has been on my mind for a while. I'm currently working on bootstrapping Consul and Vault, generating or uploading a CA into Vault, creating TLS certs for each node, and rebooting Consul and Vault with the new certs. After that, I'll look into creating an intermediate CA as well.

**Comment by siddharthist** [comment]
@j-lowry That improvement is pretty big and might take me a while. One thing I'd love to do in the meantime is to use EasyRSA locally to generate all of the certificates and push them out to the nodes. The [kubernetes role currently does something like this](https://github.com/CiscoCloud/mantl/blob/46af278aba8d5909f25f2e771ff9c9d9b399ca65/roles/kubernetes/tasks/secrets.yml), but it should be done in the certificates role instead (see #1753). 

If you'd like to make a PR to that effect, I'd be happy to review it!

**Comment by j-lowry** [comment]
I think the easiest "win" right now is to remove the CA private key in the `certificates` role at the end. As far as I have seen, once the individual host certs are signed, there is no need for the private key to exist on each host. If more certificates are needed to be signed later, the private key should still exist on the operators workstation and can be backed up.

The reason to not keep the private key is not to protect the machine that has been compromised, it is to protect the other machines in the environment as much as possible. If an attacker gains root but not the CA private key, they get all the things @siddharthist described above, but at least the other machines can still trust the certificates that have been signed by the CA. If the attacker gains root on that machine *and* gets the private key, the attacker can then issue as many trusted certificates as they want, for as long as the CA is good for. The only option would be to re-key every host with a new certificate authority. Most likely, one could make the argument the entire environment is compromised and needs to rebuilt entirely. 

@siddharthist Also, I see a lot of the work headed down the line with TLS in general, would love to help out.

**Comment by siddharthist** [comment]
@j-lowry I think you're exactly right. I'd love help transitioning to generating the certificates on the provisioning machine with EasyRSA or with bootstrapping a Consul + Vault cluster to generate them, whichever you're more comfortable with!

**Comment by j-lowry** [comment]
I like what happens in the `generate-certificates` script, but I am wondering if we can move those steps to ansible to get the root CA generated on the provisioning machine. Consul and vault need to get bootstrapped, then we can move all PKI related tasks to Vault and any client that needs to connect can use vault to rotate their keys/certs.

**Comment by siddharthist** [comment]
@j-lowry Exactly! I want to move from `generate-certificates` to EasyRSA because EasyRSA is widely used, whereas we wrote `generate-certificates` without even knowing about it. We can download EasyRSA onto the provisioning machine with Ansible.
